plugins/modules/katello_upload.py

```diff
@@ -105,36 +85,61 @@ def main():
         ),
     )
 
-    entity_dict = module.parse_params()
+    entity_dict = module.clean_params()
 
     module.connect()
 
-    entity_dict['organization'] = find_organization(module, name=entity_dict['organization'])
-    entity_dict['product'] = find_product(module, name=entity_dict['product'], organization=entity_dict['organization'])
-    entity_dict['repository'] = find_repository(module, name=entity_dict['repository'], product=entity_dict['product'])
+    entity_dict['organization'] = module.find_resource_by_name('organizations', entity_dict['organization'], thin=True)
+    scope = {'organization_id': entity_dict['organization']['id']}
+    entity_dict['product'] = module.find_resource_by_name('products', entity_dict['product'], params=scope, thin=True)
+    product_scope = {'product_id': entity_dict['product']['id']}
+    entity_dict['repository'] = module.find_resource_by_name('repositories', entity_dict['repository'], params=product_scope)
+    repository_scope = {'repository_id': entity_dict['repository']['id']}
+
+    filename = os.path.basename(entity_dict['src'])
+
+    checksum = hashlib.sha256()
+    with open(entity_dict['src'], 'rb') as contentfile:
+        for chunk in iter(lambda: contentfile.read(CONTENT_CHUNK_SIZE), b""):
+            checksum.update(chunk)
+    checksum = checksum.hexdigest()
 
     content_unit = None
-    if entity_dict['repository'].content_type == "yum":
-        name, version, release, arch = check_output("rpm --queryformat '%%{NAME} %%{VERSION} %%{RELEASE} %%{ARCH}' -qp %s" % entity_dict['src'],
-                                                    shell=True).decode('ascii').split()
-        query = "name = \"{}\" and version = \"{}\" and release = \"{}\" and arch = \"{}\"".format(name, version, release, arch)
-        content_unit = find_package(module, query, repository=entity_dict['repository'], failsafe=True)
-    elif entity_dict['repository'].content_type == "file":
-        h = hashlib.sha256()
-        with open(entity_dict['src'], "rb") as f:
-            for chunk in iter(lambda: f.read(4096), b""):
-                h.update(chunk)
-        checksum = h.hexdigest()
-        name = os.path.basename(entity_dict['src'])
-        query = "name = \"{}\" and checksum = \"{}\"".format(name, checksum)
-        content_unit = find_file(module, query, repository=entity_dict['repository'], failsafe=True)
+    if entity_dict['repository']['content_type'] == 'yum':
+        name, epoch, version, release, arch = check_output(['rpm', '--queryformat', '%{NAME} %{EPOCHNUM} %{VERSION} %{RELEASE} %{ARCH}',
```

Comment:
Could this be done with a python rpm module?
Not in this PR, but generally...

Reply:
yes, https://github.com/theforeman/foreman-ansible-modules/issues/407